Enable Support for IPv6 PXE

dacianstremtan opened the pull request:
## Description of the Pull Request (PR):

This allow PXE using IPv6 IP address for the controller. 

Manual change needs to be done to the 

/etc/warewulf/ipxe/default.pxe:
e.g. set baseuri http://{{.Ipaddr6}}:{{.Port}}/provision/{{.Hwaddr}}

/etc/dhcp/dhcpd.conf: 
....
if exists user-class and option user-class = "iPXE" {
    filename "http://[.Ipaddr6]:9873/ipxe/${mac:hexhyp}";
} else 
...
/etc/dhcp/dhcpd6.conf:
....
if exists dhcp6.user-class and substring (option dhcp6.user-class, 2, 4) = "iPXE" {
       option dhcp6.bootfile-url "http://[.Ipaddr6]:9873/ipxe/${mac:hexhyp}";
}
else if exists dhcp6.client-arch-type and option dhcp6.client-arch-type = 00:07 {
       option dhcp6.bootfile-url "tftp://[.Ipaddr6]/warewulf/snponly.efi" ;
}
....


## This fixes or addresses the following GitHub issues:

- Fixes #


## Reviewer  checklist

The reviewer checks the following items before merging the PR.

- [ ] The PR is based on the appropriate branch (typically [main](https://github.com/warewulf/warewulf/tree/main/userdocs))
- [ ] All commits are "Signed off" (e.g., using `git commit --signoff`) in agreement to the [DCO](DCO.txt)
- [ ] The [CHANGELOG](https://github.com/warewulf/warewulf/blob/main/CHANGELOG.md) has been updated, if necessary, and under the correct release heading
- [ ] The [userdocs](https://github.com/warewulf/warewulf/tree/main/userdocs) have been updated, if necessary
- [ ] The submitter is listed in the [contributors file](https://github.com/warewulf/warewulf/blob/main/CONTRIBUTORS.md)
- [ ] The test suite has been updated, if necessary

dacianstremtan commented:
> > e.g. set baseuri http://{{.Ipaddr6}}:{{.Port}}/provision/{{.Hwaddr}}
> 
> This is not a valid URI w/o the `[]` around the raw IPv6 address, why is this the case here?

Because it works. I wanted to use minimum changes to the file. Since it worked from the start I did not try a different approach.

<img width="800" height="600" alt="Screenshot_client1-rocky8_2025-07-21_10:09:53" src="https://github.com/user-attachments/assets/5e643a64-fb98-40c3-893c-61a827196658" />
<img width="800" height="600" alt="Screenshot_client1-rocky8_2025-07-21_10:10:01" src="https://github.com/user-attachments/assets/8371df8f-8197-4958-a0d2-d87075165c29" />

dacianstremtan commented:
With square brackets also works. I think we should use square brackets when for consistent configurations

Dacian
<img width="800" height="600" alt="Screenshot_client1-rocky8_2025-07-21_10:18:36" src="https://github.com/user-attachments/assets/387ace3a-a907-4c34-a8f2-679cc03c2fd6" />

middelkoopt commented on internal/pkg/config/root.go:
These changes have been dropped in #1958 in favor of more recent commits.

middelkoopt replied:
`conf.Ipaddr6` now stored with prefix as per more recent commit.

middelkoopt reviewed (changes requested):
I rebased and reworked this PR in #1958.  In order for it to be merged with #1958 you will have to add a signed-off for each commit, please do not rebase as I've already rebased it.  Alternatively I could just squash your commits and add a Co-authored-by - but will need to check if this is sufficient Signed-off-by.

Please note, that you should use `{{.Authority}}` template variable for the address/port combo with `[]` inserted for IPv6 addresses.  `.Ipaddr6` is still set as your original patch.  The `deafult.ipxe` has been upated for this.

Please test to see if this works for you.  

Please let me know if this approach works for you.

It would be great if you could send us your overlay files to help build out full IPv6 support.

middelkoopt commented on internal/pkg/config/root.go:
```diff
@@ -137,14 +137,19 @@ func (conf *WarewulfYaml) Parse(data []byte, autodetect bool) error {
-			}
-		} else {
-			return fmt.Errorf("invalid ipv6 address: must use CIDR notation: %s", conf.Ipaddr6)
+		if conf.Ipv6net == "" {
+			wwlog.Error("Ipv6 network has not been set in warewulf.conf: ipv6net")
+			return fmt.Errorf("Invalid Ipv6 network")
+		}
+		_, ipv6net, err := net.ParseCIDR(conf.Ipv6net)
+		if err != nil {
+			wwlog.Error("Invalid Ipv6 address specified, must be CIDR notation: %s", conf.Ipv6net)
+			return fmt.Errorf("Invalid Ipv6 network")
+		}
+		if msize, _ := ipv6net.Mask.Size(); msize > 64 {
+			wwlog.Error("ipv6 mask size must be smaller than 64")
+			return fmt.Errorf("Invalid Ipv6 network size")
```
This section has been removed as part of the rework patch #1958.

dacianstremtan commented:
Hi,
"Please let me know if this approach works for you."
Yes. I downloaded and compiled your the tm-ipv6-pxe branch and it works for
me.

"It would be great if you could send us your overlay files to help build
out full IPv6 support."
I don't have specific overlay files. Maybe you can expand on this request.

Dacian


În lun., 28 iul. 2025 la 16:40, Timothy Middelkoop ***@***.***>
a scris:

> ***@***.**** requested changes on this pull request.
>
> I rebased and reworked this PR in #1958
> <https://github.com/warewulf/warewulf/pull/1958>. In order for it to be
> merged with #1958 <https://github.com/warewulf/warewulf/pull/1958> you
> will have to add a signed-off for each commit, please do not rebase as I've
> already rebased it. Alternatively I could just squash your commits and add
> a Co-authored-by - but will need to check if this is sufficient
> Signed-off-by.
>
> Please note, that you should use {{.Authority}} template variable for the
> address/port combo with [] inserted for IPv6 addresses. .Ipaddr6 is still
> set as your original patch. The deafult.ipxe has been upated for this.
>
> Please test to see if this works for you.
>
> Please let me know if this approach works for you.
>
> It would be great if you could send us your overlay files to help build
> out full IPv6 support.
> ------------------------------
>
> In internal/pkg/config/root.go
> <https://github.com/warewulf/warewulf/pull/1852#discussion_r2217292585>:
>
> > +		if conf.Ipv6net == "" {
> +			wwlog.Error("Ipv6 network has not been set in warewulf.conf: ipv6net")
> +			return fmt.Errorf("Invalid Ipv6 network")
> +		}
> +		_, ipv6net, err := net.ParseCIDR(conf.Ipv6net)
> +		if err != nil {
> +			wwlog.Error("Invalid Ipv6 address specified, must be CIDR notation: %s", conf.Ipv6net)
> +			return fmt.Errorf("Invalid Ipv6 network")
> +		}
> +		if msize, _ := ipv6net.Mask.Size(); msize > 64 {
> +			wwlog.Error("ipv6 mask size must be smaller than 64")
> +			return fmt.Errorf("Invalid Ipv6 network size")
>
> Why is Ipv6net being set explicitly? Ipv6net is being set from Ipaddr6 and
> seems redundant to have two things in the configuration. The current IPv4
> code works this way now.
>
> As for mask size, for most cases having a prefix greater than 64 is a bad
> idea but it could be used in a static case, and this prevents this. At
> least make it a warning if not remove.
>
> As part of merge/testing effort I'm removing this block.
> ------------------------------
>
> On internal/pkg/config/root.go
> <https://github.com/warewulf/warewulf/pull/1852#discussion_r2237810013>:
>
> These changes have been dropped in #1958
> <https://github.com/warewulf/warewulf/pull/1958> in favor of more recent
> commits.
> ------------------------------
>
> On internal/pkg/config/root.go
> <https://github.com/warewulf/warewulf/pull/1852#discussion_r2237811993>:
>
> conf.Ipaddr6 now stored with prefix as per more recent commit.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/warewulf/warewulf/pull/1852#pullrequestreview-3035349795>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIRPEFGUWUFJD6AOB4TQLRD3K2DCJAVCNFSM6AAAAAB2KU3C7SVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMZVGM2DSNZZGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

middelkoopt commented:
Good to hear that it works.

I mean the dhcpd6.conf and other configuration files.  It is possible to generate files on the host from a template (host overlay files) that get generated with`wwctl configure`.  Did you use this or just create/edit the files yourself? 

Also along similar lines, do you use /etc/hosts for node IP addresses?  If so, how is this generated/maintained.

Could you update your commit's to have a signed-of-by line?

middelkoopt commented:
I squashed commits and added a "co-authored-by" attribution.

anderbubble commented:
Closing in favor of #1958.

dacianstremtan commented:
Thank you.
I appreciate your help in moving this forward.
I will share my dhcpv6.conf and dhcpd.conf files today
Dacian

În vin., 1 aug. 2025 la 14:42, Timothy Middelkoop ***@***.***>
a scris:

> *middelkoopt* left a comment (warewulf/warewulf#1852)
> <https://github.com/warewulf/warewulf/pull/1852#issuecomment-3145446019>
>
> I squashed commits and added a "co-authored-by" attribution.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/warewulf/warewulf/pull/1852#issuecomment-3145446019>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIRPEFC2YESLYSZFE4KFSAL3LOYJJAVCNFSM6AAAAAB2KU3C7SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCNBVGQ2DMMBRHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
